CLN: Refactor pandas/tests/base - part3

SaturnFromTitan opened the pull request:
Part of #23877

- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

-----------------------
Starting to dissolve the `Ops` Mixin to drive fixturization of `pandas/tests/base/`

SaturnFromTitan committed:
added new fixtures that can replace the Ops mixin

SaturnFromTitan commented on pandas/tests/base/utils.py:
```diff
@@ -0,0 +1,59 @@
+            err = TypeError
+
+        with pytest.raises(err):
+            getattr(obj, op)
```
These methods are basically the extracted methods of the `Ops` mixin. I need them as separte functions so I can get rid of the usage of `self.is_valid_objs` and `self.objs`.

SaturnFromTitan commented on pandas/conftest.py:
```diff
@@ -881,3 +881,75 @@ def index_or_series(request):
+
+
+@pytest.fixture(params=_all_objs, ids=_all_obj_ids)
+def index_or_series_obj(request):
```
There is already a fixture called `index_or_series` which just returns the class. I'm open to suggestions for better names.

SaturnFromTitan committed:
fixed type hinting

WillAyd commented on pandas/conftest.py:
```diff
@@ -881,3 +881,75 @@ def index_or_series(request):
     return request.param
+
+
+_int_index = tm.makeIntIndex(10, name="a")
```
Are these fixtures used outside of `base`? If not, then should just create a conftest in that directory to house

SaturnFromTitan replied:
Yes, they will be used as well in:
- `pandas/tests/indexes/datetimes`
- `pandas/tests/indexes/period`
- `pandas/tests/indexes/timedeltas`

SaturnFromTitan commented on pandas/tests/base/test_ops.py:
```diff
@@ -130,27 +130,28 @@ def check_ops_properties(self, props, filter=None, ignore_failures=False):
+
+@pytest.mark.parametrize("klass", [Series, DataFrame])
+def test_binary_ops_docs(klass):
+    op_map = {
```
I didn't want to do any refactoring in this ticket yet. This will come in a later PR

SaturnFromTitan commented:
Thanks for your reviews!

SaturnFromTitan committed:
refactoring according to review comments

SaturnFromTitan commented:
I addressed all your comments @WillAyd @jschendel

WillAyd commented on pandas/conftest.py:
```diff
@@ -881,3 +881,71 @@ def index_or_series(request):
+    return _indexes[request.param].copy()
+
+
+_arr = arr = np.random.randn(10)
```
Is `arr` used here?

SaturnFromTitan replied:
whoops, no that's just a mistake

SaturnFromTitan committed:
removing unused variable arr and fixing the _narrow_series data

jreback commented on pandas/conftest.py:
```diff
@@ -881,3 +881,72 @@ def index_or_series(request):
+def _create_narrow_series(data, data_dtype):
+    """ Helper for the _narrow_series dict """
+    index = _indexes["int-index"].copy()
+    return pd.Series(data.astype(data_dtype), index=index, name="a")
```
this needs to conform with pandas/tests/indexes/conftest.py (likely easier to pull that up to top-level)

jreback commented on pandas/conftest.py:
```diff
@@ -881,3 +881,72 @@ def index_or_series(request):
+    """
+    Fixture for tests on indexes
+
+    copy to avoid mutation, e.g. setting .name
```
for now just create fixtures in pandas/tests/base

then can move common fixtures to a higher level. *only* really shared fixtures should be in this file.

jreback commented on pandas/tests/base/utils.py:
```diff
@@ -0,0 +1,59 @@
+import pandas.util.testing as tm
+
+
+def allow_na_ops(obj: Any) -> bool:
```
what is the purpose of this file, it doesn't appear to be imported anywhere

SaturnFromTitan commented:
@jreback Like mentioned in the issue description and in [this comment](https://github.com/pandas-dev/pandas/pull/30147#discussion_r355307827): The utils and fixtures in this PR will be used in more places than just `tests/base`. That's why I put the fixtures directly into the base conftest. Applying them right away would have yielded a huge PR, so I wanted to split it into parts.

I see now that not using the utils and fixtures anywhere makes it impossible to understand their purpose. I will update the PR and remove the usage of `Ops` in `tests/indexes` which needs much fewer changes than `tests/base`.

jreback commented:
> @jreback Like mentioned in the issue description and in [this comment](https://github.com/pandas-dev/pandas/pull/30147#discussion_r355307827): The utils and fixtures in this PR will be used in more places than just `tests/base`. That's why I put the fixtures directly into the base conftest. Applying them right away would have yielded a huge PR, so I wanted to split it into parts.
> 
> I see now that not using the utils and fixtures anywhere makes it impossible to understand their purpose. I will update the PR and remove the usage of `Ops` in `tests/indexes` which needs much fewer changes than `tests/base`.

this is likely true, however, I would rather move them locally pandas/tests/base, THEN in a follow on conform names & patterns between multiple places in the codebase.

SaturnFromTitan commented:
Makes sense. It makes the PRs more atomic

SaturnFromTitan committed:
removing usage of Ops mixin in tests/indexes

SaturnFromTitan committed:
moved new fixtures to tests/indexes/conftest.py as they're not used anywhere else for now

SaturnFromTitan committed:
using new utils and fixtures in tests/indexes

SaturnFromTitan committed:
fixturizing tests/base/test_ops.py::test_binary_ops_docs

SaturnFromTitan committed:
refactoring of tests/indexes/conftest.py

SaturnFromTitan commented:
@jreback @WillAyd @jschendel 
Thanks for your feedback. I added the following changes to the PR:
1. I extended the scope of this PR and replaced the usage of the `Ops` mixin in `tests/indexes/` to provide some more context of the added fixtures and utils.
1. Like suggested, I moved the new fixtures into `tests/indexes/conftest.py` as that's the only place the fixtures are used for now
1. I'm using the existing `indeces_dict` from `tests/indexes/conftest.py` now instead of creating my own `indexes` fixtures now.

Please have another look at the PR

WillAyd commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -12,39 +12,45 @@
+    def test_invalid_ops_properties(self, op, index_or_series_obj):
+        obj = index_or_series_obj
+        if isinstance(obj, (DatetimeIndex, PeriodIndex)):
+            return
```
Minor nit but would rather `pytest.skip` than return so these don't show as green

WillAyd reviewed (commented):
Looks good one comment

SaturnFromTitan committed:
using pytest.skip instead of early return

SaturnFromTitan committed:
skipping broken tests

SaturnFromTitan committed:
replaced more return statements with pytest.skip

SaturnFromTitan commented:
Some tests in `tests/indexes/test_base.py` break with the updated `indices` fixture. At least the `union` and `intersection` cases should be fixed I guess (they break with Index.union or Index.intersection is called from a tz-aware index on an array-like). If you agree I can create issues for it.

jreback commented on pandas/tests/base/test_ops.py:
```diff
@@ -83,74 +76,29 @@ def setup_method(self, method):
+@pytest.mark.parametrize(
+    "op_name, op",
+    [
+        ("add", "+"),
```
we have all kinds of fixtures for ops *already* in pandas/conftest.py, see if you can use them

jreback replied:
so you want to use all_arithmetic_functions, then simply do
then use pandas.core.ops.__get_opstr

SaturnFromTitan replied:
Great! Thanks for the hint

jreback commented on pandas/tests/base/utils.py:
```diff
@@ -0,0 +1,32 @@
+from typing import Any
+
+import numpy as np
+
```
rather than creating this file can you put in pandas/utils/testing.py (you may need to expand the names to conform a bit more)

jreback commented on pandas/tests/indexes/conftest.py:
```diff
@@ -50,3 +51,51 @@ def zero(request):
     return request.param
+
+
+def _create_series(index):
```
can we do this as another step. I am really not in love with the way these fixtures are created, they are duplicative of many other fixtures and named in a funny way.

SaturnFromTitan replied:
I completely agree that these fixtures feel weird. However, I need them to remove the `Ops` Mixin, which is basically the goal of this PR.

...and even though they are weird, the same weirdness is going on already in the code. So it's at least not making anything worse ✌️

jreback commented on pandas/tests/indexes/period/test_ops.py:
```diff
@@ -2,24 +2,29 @@
-        self.is_valid_objs = [o for o in self.objs if mask(o)]
-        self.not_valid_objs = [o for o in self.objs if not mask(o)]
+class TestPeriodIndexOps:
+    @pytest.mark.parametrize("op", PeriodArray._datetimelike_ops)
```
i would rather actually either hardcode these ops and/or create a fixture. we already do a lot of this in pandas/tests/series/test_datetime_values.py where we test all accessors, so this is likely duplicative. (though its possible we are just not fully testing the Index ones).

I am not saying to actually fix this here, but to add appropriate TODOs / lists of things to come back around and patch.

SaturnFromTitan replied:
Would creating a follow-up ticket be enough?

SaturnFromTitan replied:
You're right, this is already tested in `pandas/tests/series/test_datetime_values.py::test_dt_namespace_accessor`. I'll remove my test cases then

jreback commented on pandas/tests/indexes/timedeltas/test_ops.py:
```diff
@@ -6,24 +6,30 @@
-        self.is_valid_objs = [o for o in self.objs if mask(o)]
-        self.not_valid_objs = []
+class TestTimedeltaIndexOps:
+    @pytest.mark.parametrize("op", TimedeltaIndex._datetimelike_ops)
```
same as above

SaturnFromTitan replied:
Same as above

jreback reviewed (changes requested):
see my comments. moving tests / around fixtures is very tricky. i would break this apart even more.

jbrockmendel commented on pandas/tests/series/conftest.py:
```diff
@@ -3,16 +3,6 @@
-    """
-    s = tm.makeTimeSeries()
-    s.name = "ts"
-    return s
```
IIRC the name of this has caused some confusion, since i think it is a numeric Series with a DatetimeIndex.  might be worth giving a more informative name

SaturnFromTitan replied:
Good point, I wouldn't want to tackle that in this PR though. I'm just moving it to a different location so I have access to it in the tests I'm changing.

jreback commented:
@SaturnFromTitan is this still active?

SaturnFromTitan commented:
Yes it is @jreback I'll continue after the holidays

WillAyd commented:
@SaturnFromTitan looks like some merge conflicts - can you address?

SaturnFromTitan commented:
Sorry for the abstinence. I'll pick this issue up this week.

SaturnFromTitan committed:
Merge branch 'master' into refactor-test-base-part3

SaturnFromTitan committed:
took care of review comments

SaturnFromTitan commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -12,32 +12,37 @@
-        f = lambda x: isinstance(x, DatetimeIndex)
-        self.check_ops_properties(DatetimeIndex._field_ops, f)
-        self.check_ops_properties(DatetimeIndex._object_ops, f)
-        self.check_ops_properties(DatetimeIndex._bool_ops, f)
```
Already tested in `pandas/tests/series/test_datetime_values.py::test_dt_namespace_accessor`

jbrockmendel replied:
Good catch.  test_dt_namespace_accessor could use a good refactor if youre up to it (separate PR)

SaturnFromTitan committed:
removed redundant tests

SaturnFromTitan commented on pandas/tests/base/test_ops.py:
```diff
@@ -83,74 +75,29 @@ def setup_method(self, method):
 
         self.objs = self.indexes + self.series + self.narrow_series
 
-    def check_ops_properties(self, props, filter=None, ignore_failures=False):
```
This method was only used in pandas/tests/indexes. I could remove all tests where it was used though as they were redundant (already tested in `pandas/tests/series/test_datetime_values.py::test_dt_namespace_accessor`)

SaturnFromTitan committed:
removed changes that aren't needed for this PR anymore

SaturnFromTitan committed:
moved helper to more appropriate location

SaturnFromTitan committed:
Merge branch 'master' into refactor-test-base-part3

SaturnFromTitan committed:
removed some outdated changes

SaturnFromTitan commented on pandas/tests/indexes/period/test_ops.py:
```diff
@@ -2,25 +2,11 @@
-        self.is_valid_objs = [o for o in self.objs if mask(o)]
-        self.not_valid_objs = [o for o in self.objs if not mask(o)]
-
-    def test_ops_properties(self):
```
Already tested in `pandas/tests/series/test_datetime_values.py::test_dt_namespace_accessor`

SaturnFromTitan commented on pandas/tests/indexes/timedeltas/test_ops.py:
```diff
@@ -8,26 +8,13 @@
-        self.is_valid_objs = [o for o in self.objs if mask(o)]
-        self.not_valid_objs = []
-
-    def test_ops_properties(self):
```
Already tested in `pandas/tests/series/test_datetime_values.py::test_dt_namespace_accessor`

SaturnFromTitan committed:
moved datetime_series fixture to the root conftest so it's available for index tests

SaturnFromTitan commented:
@jreback Thanks to your comment about redundant tests I could reduce the scope of this PR quite a bit further. It's ready to review now.

SaturnFromTitan commented on pandas/conftest.py:
```diff
@@ -868,6 +868,16 @@ def tick_classes(request):
 
 
+@pytest.fixture
+def datetime_series():
```
Moved from `pandas/tests/series/conftest.py`

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -298,7 +298,7 @@ def test_ensure_copied_data(self, indices):
             # RangeIndex cannot be initialized from data
             # MultiIndex and CategoricalIndex are tested separately
-            return
+            pytest.skip()
```
can you add a message to the skip

SaturnFromTitan replied:
I'll just remove them from this PR, as it's not really connected. It was rather a left-over from the initially suggested changes

jreback reviewed (changes requested):
ok looks pretty good. @jbrockmendel if you'd have a look

SaturnFromTitan committed:
using all_arithmetic_functions in test_binary_ops_docs now

SaturnFromTitan committed:
undid some unrelated changes

jbrockmendel commented on pandas/core/ops/__init__.py:
```diff
@@ -265,7 +265,7 @@ def _get_opstr(op):
         operator.floordiv: "//",
         rfloordiv: "//",
-        operator.mod: None,  # TODO: Why None for mod but '%' for rmod?
+        operator.mod: "%",
```
do we have tests where this makes a difference?  maybe this lets us use numexpr and performance is affected?

SaturnFromTitan replied:
Yeah, I wasn't sure about this one. Thought I give it a try and see if any of the CI hooks is failing. The 2 failing ones in `pandas-dev.pandas (Linux py37_locale)` look unrelated though. I'll rebase and try again...

SaturnFromTitan replied:
I undid the change now and handle the symbol conversion in the test case itself. The CI is still failing though...

SaturnFromTitan replied:
@jbrockmendel CI is green now, so all tests seem to be fine with this change.
@jreback do you have any concerns about it?

jreback replied:
looks ok, original issue is that this looks like a format string to python

jbrockmendel commented on pandas/tests/base/test_ops.py:
```diff
@@ -83,74 +83,23 @@ def setup_method(self, method):
 
-    def check_ops_properties(self, props, filter=None, ignore_failures=False):
-        for op in props:
-            for o in self.is_valid_objs:
```
im very happy to see this pattern go

SaturnFromTitan committed:
Merge branch 'master' into refactor-test-base-part3

SaturnFromTitan committed:
undid my change in pandas/core/ops/__init__.py and handle the case differently in pandas/tests/base/test_ops.py

SaturnFromTitan committed:
Revert "undid my change in pandas/core/ops/__init__.py and handle the case differently in pandas/tests/base/test_ops.py"

This reverts commit 8bf1142d7f0790288071f450903ff5d516618c09.

SaturnFromTitan committed:
Merge branch 'master' into refactor-test-base-part3

SaturnFromTitan commented:
Friendly reminder @jreback

jbrockmendel commented on pandas/tests/base/test_ops.py:
```diff
@@ -83,74 +83,23 @@ def setup_method(self, method):
+def test_binary_ops_docs(klass, all_arithmetic_functions):
+    operator = all_arithmetic_functions
+    operator_name = _get_op_name(operator, special=False)
+    operator_symbol = _get_opstr(operator)
```
i think we might want to keep the hard-coded versions we have below rather than re-use this, otherwise a bug in this function wouldn't be caught?

SaturnFromTitan replied:
That's what I had first, but @jreback suggested to use the `all_arithmetic_functions` fixture combined with `_get_opstr` instead.

I see your point now though. The `_get_opstr` method itself is used in the implementation of the docstring?

So I'll revert it again

SaturnFromTitan committed:
using hard-coded values to parametrize test_binary_ops

jreback commented:
thanks @SaturnFromTitan yeah love these test refactor PRs; we have built up some monotliths...though many are being taken down / split now, so thanks.
